### ISSUE TITLE
Some combination of structs and functions causes compiler infinite loop

### ISSUE THREAD

**Issue by onovytskyi** [microsoft/DirectXShaderCompiler#2514]
Hello,

Some combination of structs and functions leads to infinite loop on the latest released compiler,
older compiler versions crash with the same code.
I managed to distill out the code that causes the issue to this
```
struct Resources
{
    float a; // removing this float removes the issue
    float b;
};

struct ResourceContainer
{
    Resources resources;
};

struct Data
{
    float constants; // removing this float removes the issue
    ResourceContainer container;

    ResourceContainer GetContainer()
    {
        ResourceContainer i;
        i.resources = container.resources;
        return i;
    }
};

Data CreateData()
{
    Data i;
    return i;
}

static const Data data = CreateData();
static const ResourceContainer container = data.GetContainer();

float4 main() : SV_Target
{
    return (float4)0;
}
```
Cheers,
Oleksandr

**Comment by BitMD** [comment]
Thanks for the clear repro.  We'll take a look.

**Comment by python3kgae** [comment]
Fixed with #2519
